File: grind75_python/floodFill.py
# https://leetcode.com/problems/flood-fill/description/

class Solution:

  def valid(self, r, c, image):
    if 0 <= r < len(image) and 0 <= c < len(image[0]):
      return True
    return False

  def dfs(self, r, c, visited, image, color):
    visited.add((r, c))
    prevColor = image[r][c]
    image[r][c] = color
    if (r+1, c) not in visited and self.valid(r+1, c, image) and image[r+1][c] == prevColor:
      self.dfs(r+1, c, visited, image, color)
    if (r-1, c) not in visited and self.valid(r-1, c, image) and image[r-1][c] == prevColor:
      self.dfs(r-1, c, visited, image, color)
    if (r, c-1) not in visited and self.valid(r, c-1, image) and image[r][c-1] == prevColor:
      self.dfs(r, c-1, visited, image, color)
    if (r, c+1) not in visited and self.valid(r, c+1, image) and image[r][c+1] == prevColor:
      self.dfs(r, c+1, visited, image, color)

  def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
    visited = set()
    self.dfs(sr, sc, visited, image, color)
    return image

File: CPP/floodFill.cpp
// https://leetcode.com/problems/flood-fill/

#include <bits/stdc++.h>
using namespace std;

class Solution {
 public:
  void dfs(vector<vector<int>>& image, int r, int c, int color, int newColor) {
    if (image[r][c] == color) {
      image[r][c] = newColor;
      if (r >= 1) dfs(image, r - 1, c, color, newColor);
      if (c >= 1) dfs(image, r, c - 1, color, newColor);
      if (r + 1 < image.size()) dfs(image, r + 1, c, color, newColor);
      if (c + 1 < image[0].size()) dfs(image, r, c + 1, color, newColor);
    }
  }
  vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
    int ogColor = image[sr][sc];
    if (ogColor != color) dfs(image, sr, sc, ogColor, color);
    return image;
  }
};